### PR TITLE
separate upstreamable GPIO driver from downstream GPIO macros

### PR DESCRIPTION
**Pull request by pdp7**
In preparation of RFC for the GPIO driver, I am separating the GPIO driver patch from downstream GPIO header with hacky macros used in `fuction i2c_dw_configure_gpio()` in `drivers/i2c/busses/i2c-designware-master.c`.

I plan to focus on pinctrl driver after GPIO driver is reviewed.  Once pinctrl driver is functional, the gpio macro hacks in `drivers/i2c/busses/i2c-designware-master.c` can hopefully be replaced.

### CODE REVIEW

**Comment by pdp7** [comment]
note: https://github.com/esmil/linux/pull/25/commits/132e529dbfb44a39a1a8cadb70f69bc6854e8af6 has duplicate SoB.  I will fix this up in the RFC.

The RFC will also include the gpio bindings commit https://github.com/esmil/linux/commit/d1d30dda8567322b72f75e153feb70963e4de1af in a two patch RFC series.

**Comment by pdp7** [comment]
Oops, just noticed that I forgot to put `Huan Feng <huan.feng@starfivetech.com>` in MAINTAINERS too.  Anyways, I will see what Michael and Huan say before I change it.

**Review by esmil — COMMENTED** [review]
Hmm.. something is still not right here, I think. All the `sf_vic_`-functions seem to be only reachable and used as exported functions defined in the header you just removed, and they seem to be mostly duplicate functionality from the gpio driver functions.

You might want to remove those from the driver as well before sending it upstream.

File: drivers/gpio/gpio-starfive-jh7100.c [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+	writel_relaxed(1, chip->base + GPIO_EN);
+
+	dev_info(dev, "StarFive GPIO chip registered %d GPIOs\n", ngpio);
+
+	return 0;
+}
+
+static const struct of_device_id starfive_gpio_match[] = {
+	{ .compatible = "starfive,jh7100-gpio", },
+	{ },
```
Comment (esmil) [inline drivers/gpio/gpio-starfive-jh7100.c]: nit: if you're changing this anyway i like the
`{ /* sentinel */ },`-pattern.

File: drivers/gpio/gpio-starfive-jh7100.c [hunk 2]
```diff
@@ -0,0 +1,545 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * GPIO driver for StarFive JH7100 SoC
+ *
+ * Copyright (C) 2020 Shanghai StarFive Technology Co., Ltd.
+ */
+
+#include <linux/module.h>
+#include <linux/gpio/driver.h>
```
Comment (esmil) [inline drivers/gpio/gpio-starfive-jh7100.c]: It seems we need `#include <linux/interrupt.h>` for compile-testing on other architectures.

**Comment by pdp7** [comment]
there have been a lot of changes so I don't think very useful any more

**Pull request closed without merging**